File: backend/app/api/v1/integrations.py
"""OAuth connect / callback / disconnect endpoints.

Three platforms currently supported:

- ``zoom``      — Zoom meetings + cloud recordings
- ``microsoft`` — Teams + Outlook + Calendar (one OAuth app)
- ``google``    — Google Calendar + Meet

The ``POST /connect`` endpoint returns an ``authorization_url`` that the SPA
redirects the browser to. ``GET /callback`` is called by the OAuth provider
after consent; it stores tokens and 302s back to ``FRONTEND_URL/integrations``.
``DELETE /disconnect`` soft-deletes the credential row.

Bot scheduling endpoints that used to live here were moved to
``/api/v1/internal/bot/*`` — the frontend talks to Supabase directly for
``meeting_bot_sessions`` CRUD and fires Inngest events from the browser.
"""

from __future__ import annotations

from typing import cast
from uuid import UUID

import structlog
from fastapi import APIRouter, Depends, HTTPException, Query, status
from fastapi.responses import RedirectResponse
from sqlalchemy import select
from sqlalchemy.orm import Session

from app.core.config import settings
from app.db.deps import get_db
from app.db.models import OrgMembership
from app.dependencies import AuthUser, get_current_user
from app.integrations.google import oauth as google_oauth
from app.integrations.microsoft import oauth as microsoft_oauth
from app.integrations.zoom import oauth as zoom_oauth
from app.services.oauth_tokens import (
    Platform,
    build_state,
    deactivate_credentials,
    list_user_integrations,
    redirect_uri_for,
    save_credentials,
    verify_state,
)

logger = structlog.get_logger(__name__)

router = APIRouter()


SUPPORTED_PLATFORMS: set[Platform] = {"zoom", "microsoft", "google"}


def _assert_supported(platform: str) -> Platform:
    if platform not in SUPPORTED_PLATFORMS:
        raise HTTPException(
            status_code=status.HTTP_400_BAD_REQUEST,
            detail=f"Unsupported platform '{platform}'",
        )
    return cast("Platform", platform)


def _resolve_default_org(session: Session, user_id: UUID) -> UUID:
    """Pick the user's first org membership as the default scope for the
    credential row. Users with multiple orgs are out of scope for v1 — they'll
    get tokens attached to whichever org comes back first."""
    org_id = session.scalar(
        select(OrgMembership.org_id)
        .where(OrgMembership.user_id == str(user_id))
        .limit(1)
    )
    if not org_id:
        raise HTTPException(
            status_code=status.HTTP_400_BAD_REQUEST,
            detail="User has no org membership; cannot connect integration",
        )
    return UUID(org_id)


# ---------------------------------------------------------------------------
# GET /api/v1/integrations
# ---------------------------------------------------------------------------


@router.get("")
async def list_integrations(
    user: AuthUser = Depends(get_current_user),
    session: Session = Depends(get_db),
) -> list[dict]:
    return list_user_integrations(session, user_id=user.id)


# ---------------------------------------------------------------------------
# POST /api/v1/integrations/{platform}/connect
# ---------------------------------------------------------------------------


@router.post("/{platform}/connect")
async def initiate_oauth(
    platform: str,
    user: AuthUser = Depends(get_current_user),
    session: Session = Depends(get_db),
) -> dict:
    p = _assert_supported(platform)
    org_id = _resolve_default_org(session, user.id)
    state = build_state(org_id, user.id, p)
    redirect_uri = redirect_uri_for(p)

    if p == "zoom":
        if not settings.zoom_client_id:
            raise HTTPException(500, "Zoom OAuth is not configured")
        url = zoom_oauth.build_authorize_url(
            client_id=settings.zoom_client_id,
            redirect_uri=redirect_uri,
            state=state,
        )
    elif p == "microsoft":
        if not settings.microsoft_client_id:
            raise HTTPException(500, "Microsoft OAuth is not configured")
        url = microsoft_oauth.build_authorize_url(
            client_id=settings.microsoft_client_id,
            redirect_uri=redirect_uri,
            state=state,
        )
    elif p == "google":
        if not settings.google_client_id:
            raise HTTPException(500, "Google OAuth is not configured")
        url = google_oauth.build_authorize_url(
            client_id=settings.google_client_id,
            redirect_uri=redirect_uri,
            state=state,
        )
    else:  # pragma: no cover - guarded above
        raise HTTPException(400, f"Unsupported platform {p}")

    logger.info("oauth_connect_initiated", platform=p, user_id=str(user.id))
    return {"authorization_url": url}


# ---------------------------------------------------------------------------
# GET /api/v1/integrations/{platform}/callback
# ---------------------------------------------------------------------------


@router.get("/{platform}/callback")
async def oauth_callback(
    platform: str,
    code: str | None = Query(default=None),
    state: str | None = Query(default=None),
    error: str | None = Query(default=None),
    session: Session = Depends(get_db),
) -> RedirectResponse:
    p = _assert_supported(platform)
    frontend = (settings.frontend_url or "http://localhost:3000").rstrip("/")
    return_to = f"{frontend}/integrations"

    if error:
        logger.warning("oauth_callback_provider_error", platform=p, error=error)
        return RedirectResponse(
            url=f"{return_to}?error={error}", status_code=302
        )
    if not code or not state:
        raise HTTPException(400, "Missing code or state in OAuth callback")

    try:
        claims = verify_state(state)
    except ValueError as exc:
        raise HTTPException(400, str(exc)) from exc
    if claims.get("platform") != p:
        raise HTTPException(400, "OAuth state platform mismatch")

    org_id = UUID(claims["org_id"])
    user_id = UUID(claims["user_id"])
    redirect_uri = redirect_uri_for(p)

    if p == "zoom":
        tokens = await zoom_oauth.exchange_code(
            client_id=settings.zoom_client_id,
            client_secret=settings.zoom_client_secret,
            redirect_uri=redirect_uri,
            code=code,
        )
    elif p == "microsoft":
        tokens = await microsoft_oauth.exchange_code(
            client_id=settings.microsoft_client_id,
            client_secret=settings.microsoft_client_secret,
            redirect_uri=redirect_uri,
            code=code,
        )
    elif p == "google":
        tokens = await google_oauth.exchange_code(
            client_id=settings.google_client_id,
            client_secret=settings.google_client_secret,
            redirect_uri=redirect_uri,
            code=code,
        )
    else:  # pragma: no cover
        raise HTTPException(400, f"Unsupported platform {p}")

    save_credentials(
        session,
        org_id=org_id,
        user_id=user_id,
        platform=p,
        access_token=tokens["access_token"],
        refresh_token=tokens.get("refresh_token"),
        expires_in_seconds=tokens.get("expires_in"),
        scopes=tokens.get("scope"),
    )

    return RedirectResponse(url=f"{return_to}?connected={p}", status_code=302)


# ---------------------------------------------------------------------------
# DELETE /api/v1/integrations/{platform}/disconnect
# ---------------------------------------------------------------------------


@router.delete("/{platform}/disconnect", status_code=204)
async def disconnect_integration(
    platform: str,
    user: AuthUser = Depends(get_current_user),
    session: Session = Depends(get_db),
) -> None:
    p = _assert_supported(platform)
    org_id = _resolve_default_org(session, user.id)
    deactivate_credentials(session, org_id=org_id, user_id=user.id, platform=p)
    logger.info("oauth_disconnected", platform=p, user_id=str(user.id))

File: worker/internal/httpapi/integrations.go
package httpapi

import (
	"errors"
	"net/http"
	"net/url"
	"strings"

	"github.com/Yrock1x/AI-Notetaker-Finance/worker/internal/crypto/fernet"
	"github.com/Yrock1x/AI-Notetaker-Finance/worker/internal/integrations/oauth"
	"github.com/Yrock1x/AI-Notetaker-Finance/worker/internal/store"
	"github.com/go-chi/chi/v5"
)

// integrations.go ports app/api/v1/integrations.py: the OAuth connect / callback
// / disconnect flow for the calendar/recording providers (Google, Microsoft,
// Zoom). connect returns an authorization_url the SPA redirects to; the provider
// then hits the PUBLIC callback (no session) which exchanges the code, stores the
// Fernet-encrypted tokens, and 302s back to the frontend integrations page.

var supportedIntegrationPlatforms = map[string]bool{"zoom": true, "microsoft": true, "google": true}

// RegisterIntegrations mounts the authed list/connect/disconnect routes.
func (s *Server) RegisterIntegrations(r chi.Router) {
	r.Get("/integrations", s.integrationsList)
	r.Post("/integrations/{platform}/connect", s.integrationsConnect)
	r.Delete("/integrations/{platform}/disconnect", s.integrationsDisconnect)
}

// RegisterIntegrationsCallback mounts the PUBLIC OAuth callback (the provider
// redirects the browser here with no session cookie). Wired outside requireAuth.
func (s *Server) RegisterIntegrationsCallback(r chi.Router) {
	r.Get("/integrations/{platform}/callback", s.integrationsCallback)
}

func titleCase(s string) string {
	if s == "" {
		return s
	}
	return strings.ToUpper(s[:1]) + s[1:]
}

func (s *Server) integrationClientCreds(platform string) (id, secret string) {
	switch platform {
	case "google":
		return s.Cfg.GoogleClientID, s.Cfg.GoogleClientSecret
	case "microsoft":
		return s.Cfg.MicrosoftClientID, s.Cfg.MicrosoftClientSecret
	case "zoom":
		return s.Cfg.ZoomClientID, s.Cfg.ZoomClientSecret
	}
	return "", ""
}

func (s *Server) integrationRedirectURI(platform string) string {
	return strings.TrimRight(s.Cfg.PublicAPIURL, "/") + "/api/v1/integrations/" + platform + "/callback"
}

// GET /api/v1/integrations — the user's active integrations (is_active filtered
// by the frontend, so the shape must carry is_active).
func (s *Server) integrationsList(w http.ResponseWriter, r *http.Request) {
	u := authUserFromCtx(r.Context())
	rows, err := store.ListUserIntegrations(r.Context(), s.DB, u.ID)
	if storeError(w, err) {
		return
	}
	writeJSON(w, http.StatusOK, rows)
}

// POST /api/v1/integrations/{platform}/connect — return the authorize URL.
func (s *Server) integrationsConnect(w http.ResponseWriter, r *http.Request) {
	platform := chi.URLParam(r, "platform")
	if !supportedIntegrationPlatforms[platform] {
		writeError(w, http.StatusBadRequest, "Unsupported platform '"+platform+"'")
		return
	}
	u := authUserFromCtx(r.Context())
	orgID, err := store.DefaultOrgForUser(r.Context(), s.DB, u.ID)
	if errors.Is(err, store.ErrNotFound) {
		writeError(w, http.StatusBadRequest, "User has no org membership; cannot connect integration")
		return
	}
	if storeError(w, err) {
		return
	}
	clientID, _ := s.integrationClientCreds(platform)
	if clientID == "" {
		writeError(w, http.StatusInternalServerError, titleCase(platform)+" OAuth is not configured")
		return
	}
	state, err := store.BuildOAuthState(s.Cfg.OAuthStateSecret(), orgID, u.ID, platform)
	if err != nil {
		writeError(w, http.StatusInternalServerError, "internal error")
		return
	}
	prov, _ := oauth.ByName(platform)
	authURL := prov.BuildAuthorizeURL(clientID, s.integrationRedirectURI(platform), state)
	writeJSON(w, http.StatusOK, map[string]string{"authorization_url": authURL})
}

// GET /api/v1/integrations/{platform}/callback — PUBLIC. Exchange + store + 302.
func (s *Server) integrationsCallback(w http.ResponseWriter, r *http.Request) {
	platform := chi.URLParam(r, "platform")
	returnTo := strings.TrimRight(s.Cfg.FrontendURL, "/") + "/integrations"
	if !supportedIntegrationPlatforms[platform] {
		writeError(w, http.StatusBadRequest, "Unsupported platform '"+platform+"'")
		return
	}
	q := r.URL.Query()
	if e := q.Get("error"); e != "" {
		http.Redirect(w, r, returnTo+"?error="+url.QueryEscape(e), http.StatusFound)
		return
	}
	code, state := q.Get("code"), q.Get("state")
	if code == "" || state == "" {
		writeError(w, http.StatusBadRequest, "Missing code or state in OAuth callback")
		return
	}
	claims, err := store.VerifyOAuthState(s.Cfg.OAuthStateSecret(), state)
	if err != nil {
		writeError(w, http.StatusBadRequest, "invalid OAuth state")
		return
	}
	if claims.Platform != platform {
		writeError(w, http.StatusBadRequest, "OAuth state platform mismatch")
		return
	}

	clientID, clientSecret := s.integrationClientCreds(platform)
	prov, _ := oauth.ByName(platform)
	tokens, err := prov.ExchangeCode(r.Context(), oauthHTTPClient, clientID, clientSecret, s.integrationRedirectURI(platform), code)
	if err != nil {
		http.Redirect(w, r, returnTo+"?error=exchange_failed", http.StatusFound)
		return
	}
	fkey, err := fernet.ParseKey(s.Cfg.TokenEncryptionKey)
	if err != nil {
		writeError(w, http.StatusInternalServerError, "token encryption is not configured")
		return
	}
	if err := store.SaveCredentials(r.Context(), s.DB, fkey, store.CredentialInput{
		OrgID: claims.OrgID, UserID: claims.UserID, Platform: platform,
		AccessToken: tokens.AccessToken, RefreshToken: tokens.RefreshToken,
		ExpiresInSeconds: tokens.ExpiresIn, Scopes: tokens.Scope,
	}); err != nil {
		writeError(w, http.StatusInternalServerError, "failed to save credentials")
		return
	}
	http.Redirect(w, r, returnTo+"?connected="+platform, http.StatusFound)
}

// DELETE /api/v1/integrations/{platform}/disconnect — soft-delete the credential.
func (s *Server) integrationsDisconnect(w http.ResponseWriter, r *http.Request) {
	platform := chi.URLParam(r, "platform")
	if !supportedIntegrationPlatforms[platform] {
		writeError(w, http.StatusBadRequest, "Unsupported platform '"+platform+"'")
		return
	}
	u := authUserFromCtx(r.Context())
	orgID, err := store.DefaultOrgForUser(r.Context(), s.DB, u.ID)
	if errors.Is(err, store.ErrNotFound) {
		writeError(w, http.StatusBadRequest, "User has no org membership; cannot connect integration")
		return
	}
	if storeError(w, err) {
		return
	}
	if storeError(w, store.DeactivateCredentials(r.Context(), s.DB, orgID, u.ID, platform)) {
		return
	}
	w.WriteHeader(http.StatusNoContent)
}
